README
Roundcube Webmail (http://roundcube.net)


Introduction:
-------------
Roundcube Webmail is a browser-based multilingual IMAP client with an
application-like user interface. It provides full functionality you expect
from an e-mail client, including MIME support, address book, folder management,
message searching and spell checking. RoundCube Webmail is written in PHP and
requires the MySQL, PostgreSQL or SQLite database. The user interface is fully
skinnable using XHTML and CSS 2.

This project is meant to be a modern webmail solution which is easy to
install/configure and that runs on a standard PHP plus MySQL, PostgreSQL or SQLite
configuration. It includes open-source classes/libraries like PEAR
(http://pear.php.net) and the IMAP wrapper from IlohaMail
(http://www.ilohamail.org).

The current development skin uses icons designed by Stephen Horlander and Kevin 
Gerich for Mozilla.org.


Installation:
-------------
For detailed instructions on how to install RoundCube webmail on your server,
please refer to the INSTALL document in the same directory as this document.
Please carefully read the REQUIREMENTS section of the INSTALL instructions.


Licensing:
----------
This product is distributed under the GPL. Please read through the file
LICENSE for more information about our license.


Contribution:
-------------
Want to help make Roundcube the best webmail solution ever?
RoundCube is open source software. Our developers and contributors all
are volunteers and we're always looking for new additions and resources.
For more information visit http://roundcube.net/contribute


Contact:
--------
For any bug reports or feature requests please refer to the tracking system
at trac.roundcube.net (http://trac.roundcube.net/wiki/Howto_ReportIssues) or 
subscribe to our mailing list. See http://roundcube.net/support
for details.

You're always welcome to send a message to the project admin:
hello@roundcube.net


debian/changelog
roundcube (1.2.3+dfsg.1-1~ppa1404+1) trusty; urgency=medium

  * [64c476d] Depend against PHP5
  * [799f4e3] Disable special handling for dh_phpcomposer
  * [7026f0c] Remove dep that has been introduced for dir_to_symlink
  * Rebuild for Ubuntu trusty.

 -- Markus Frosch <lazyfrosch@debian.org>  Mon, 09 Jan 2017 16:27:19 +0100

roundcube (1.2.3+dfsg.1-1) unstable; urgency=high

  [ Guilhem Moulin ]
  * New upstream release (closes: #847287).

 -- Sandro Knauß <hefee@debian.org>  Fri, 16 Dec 2016 12:17:54 +0100

roundcube (1.2.2+dfsg.1-1) unstable; urgency=medium

  [ Sandro Knauß ]
  * Change own mailadress to new debian one
  * Update Copyright years for myself

  [ Guilhem Moulin ]
  * New upstream release.
  * debian/control:
    + Remove php-pspell from roundcube-plugins' Depends: field, as pspell is
      not used by any plugin.  However we keep it in roundcube-core's
      Recommends: field as $config['spellcheck_engine'] defaults to 'pspell'.
      (Closes: #825500).
    + Add php-crypt-gpg, php-net-sieve and php-zip to roundcube-plugins'
      Suggests: field, respectively for the enigma, managesieve, and
      zipdownload plugins.  Thanks, Jan Gerber.  (Closes: #836909).
    + Upgrade Homepage: field from http:// to https://.
  * debian/watch: fix syntax error due to missing comment markers.

 -- Sandro Knauß <hefee@debian.org>  Sun, 16 Oct 2016 18:25:05 +0200

roundcube (1.2.1+dfsg.1-2) unstable; urgency=medium

  * Update B-D according to composer.json-dist (Closes: #833461)
  * Move gpg_crypt from depends to suggests

 -- Sandro Knauß <bugs@sandroknauss.de>  Fri, 05 Aug 2016 11:20:14 +0200

roundcube (1.2.1+dfsg.1-1) unstable; urgency=medium

  [ Guilhem Moulin ]
  * d/roundcube-core.cron.d: install new cronjob (executed as www-data) to
    finally remove all records that are marked as deleted (closes: #824676).

  [ Sandro Knauß ]
  * New upstream release.
  * update update_composer.patch with upstream dependencies.
  * Use secured links

 -- Sandro Knauß <bugs@sandroknauss.de>  Wed, 03 Aug 2016 16:18:36 +0200

roundcube (1.2.0+dfsg.1-1) unstable; urgency=medium

  * New upstream release. 
  * update patches (only file hunks updates)
  * add sql updates for 1.2.0
  * make roundcube ready for PHP 7 (Closes: #821646)
    - replace dependencies from php5-* to php-* packages
  * Update default config file to match debian default installation
    - Disable spellchecking by default (needs pspell, that is only recommended)
    - Remove plugins from default config file (needs roundcube-plugins)
  * Update the internal copies of external libraries

 -- Sandro Knauß <bugs@sandroknauss.de>  Tue, 24 May 2016 21:35:06 +0200

roundcube (1.1.5+dfsg.1-1) unstable; urgency=medium

  [ Guilhem Moulin ]
  * New upstream release (Closes: #822333).
  * debian/watch:
    + Increase uscan version from 3 to 4.
    + Change release URL from http://sf.net to https://github.com .
  * debian/control:
    + Bump Standards-Version to 3.9.8 (no changes necessary).

  [ Sandro Knauß ]
  * Use and check signed tarballs
  * update patch hunks

 -- Sandro Knauß <bugs@sandroknauss.de>  Thu, 28 Apr 2016 00:25:04 +0200

roundcube (1.1.4+dfsg.1-3) unstable; urgency=medium

  * No dependency to packages that are part of roundcube itself
  * Updated composer.json so it matches debian available versions.
    (Closes: 817792)

 -- Sandro Knauß <bugs@sandroknauss.de>  Thu, 10 Mar 2016 17:44:23 +0100

roundcube (1.1.4+dfsg.1-2) unstable; urgency=medium

  [ Vincent Bernat ]
  * Use an empty array for plugin configuration templates to ensure
    Roundcube knows the configuration file is valid. Closes: #809769.
  * Pre-Depends on appriopriate dpkg version for dir_to_symlink. Related
    to #810980.

  [ Sandro Knauß ]
  * Use dh_phpcomposer to track php dependencies. (Closes: #814664)
  * Use safe urls for VCS fields
  * Bumped compat level to 9
  * Updated lintian overrides
  * Added php-pear to depends for roundcube-core. (Closes: #801973)
  * Bump Standards-Version to 3.9.7

 -- Sandro Knauß <bugs@sandroknauss.de>  Wed, 24 Feb 2016 15:17:35 +0100

roundcube (1.1.4+dfsg.1-1) unstable; urgency=medium

  * New upstream release.

 -- Guilhem Moulin <guilhem@guilhem.org>  Mon, 28 Dec 2015 15:22:07 +0100

roundcube (1.1.3+dfsg.1-1) unstable; urgency=medium

  [ Guilhem Moulin ]
  * New upstream release.
  * Add self to Uploaders.
  * debian/copyright: Add program/js/tinymce/plugins/media/moxieplayer.swf
    to Files-Excluded.
  * Remove unused lintian overide:
    package-contains-broken-symlink var/lib/roundcube/config/config.inc.php
    etc/roundcube/config.inc.php .

 -- Guilhem Moulin <guilhem@guilhem.org>  Sat, 17 Oct 2015 23:28:44 +0200

roundcube (1.1.2+dfsg.1-5) unstable; urgency=medium

  [ Vincent Bernat ]
  * Demote php-net-ldap3 to Recommends.

  [ Guilhem Moulin ]
  * Roundcube: Add 'Breaks: roundcube-core (<< 1.1.1+dfsg.1-1)' to enable
    smooth upgrade from Wheezy and its backports.  Closes: #800659.

 -- Vincent Bernat <bernat@debian.org>  Tue, 06 Oct 2015 15:17:03 +0200

roundcube (1.1.2+dfsg.1-4) unstable; urgency=medium

  * Also recommends php5-fpm and spawn-fcgi since they don't provide
    httpd-cgi. Closes: #731705.
  * Replace /var/lib/roundcube/config with a symlink. Closes: #629032.
  * Require php-net-ldap3 for LDAP. Closes: #785056.
  * Recommends php-net-sieve. Closes: #798307.

 -- Vincent Bernat <bernat@debian.org>  Fri, 11 Sep 2015 08:28:03 +0200

roundcube (1.1.2+dfsg.1-3) unstable; urgency=medium

  * Handle dir to symlink migration for /usr/share/doc/roundcube{,-plugins}.
    Closes: #788448
  * Put a warning about access rights in empty configuration files for
    plugins. Closes: #739592
  * Depends on php5-pspell for roundcube-plugins. Closes: #793857
  * Remove /etc/roundcube/config.inc.php on purge. Closes: #793858
  * Add back Enigma plugin. Closes: #771659
  * Redact logout.html to not use a remote jquery.js.
  * Don't ship license.txt with TinyMCE.


index.php
<?php
/*
 +-------------------------------------------------------------------------+
 | Roundcube Webmail IMAP Client                                           |
 | Version 0.5.3                                                           |
 |                                                                         |
 | Copyright (C) 2005-2011, Roundcube Dev. - Switzerland                   |
 |                                                                         |
 | This program is free software; you can redistribute it and/or modify    |
 | it under the terms of the GNU General Public License version 2          |
 | as published by the Free Software Foundation.                           |
 |                                                                         |
 | This program is distributed in the hope that it will be useful,         |
 | but WITHOUT ANY WARRANTY; without even the implied warranty of          |
 | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           |
 | GNU General Public License for more details.                            |
 |                                                                         |
 | You should have received a copy of the GNU General Public License along |
 | with this program; if not, write to the Free Software Foundation, Inc., |
 | 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.             |
 |                                                                         |
 +-------------------------------------------------------------------------+
 | Author: Thomas Bruederli <roundcube@gmail.com>                          |
 +-------------------------------------------------------------------------+

 $Id: index.php 4831 2011-06-02 13:36:57Z thomasb $

*/

// include environment
require_once 'program/include/iniset.php';

// init application, start session, init output class, etc.
$RCMAIL = rcmail::get_instance();

// turn on output buffering
ob_start();

// check if config files had errors
if ($err_str = $RCMAIL->config->get_error()) {
  raise_error(array(
    'code' => 601,
    'type' => 'php',
    'message' => $err_str), false, true);
}

// check DB connections and exit on failure
if ($err_str = $DB->is_error()) {
  raise_error(array(
    'code' => 603,
    'type' => 'db',
    'message' => $err_str), FALSE, TRUE);
}

// error steps
if ($RCMAIL->action=='error' && !empty($_GET['_code'])) {
  raise_error(array('code' => hexdec($_GET['_code'])), FALSE, TRUE);
}

// check if https is required (for login) and redirect if necessary
if (empty($_SESSION['user_id']) && ($force_https = $RCMAIL->config->get('force_https', false))) {
  $https_port = is_bool($force_https) ? 443 : $force_https;
  if (!rcube_https_check($https_port)) {
    $host  = preg_replace('/:[0-9]+$/', '', $_SERVER['HTTP_HOST']);
    $host .= ($https_port != 443 ? ':' . $https_port : '');
    header('Location: https://' . $host . $_SERVER['REQUEST_URI']);
    exit;
  }
}

// trigger startup plugin hook
$startup = $RCMAIL->plugins->exec_hook('startup', array('task' => $RCMAIL->task, 'action' => $RCMAIL->action));
$RCMAIL->set_task($startup['task']);
$RCMAIL->action = $startup['action'];

// try to log in
if ($RCMAIL->task == 'login' && $RCMAIL->action == 'login') {
  $request_valid = $_SESSION['temp'] && $RCMAIL->check_request(RCUBE_INPUT_POST, 'login');

  // purge the session in case of new login when a session already exists 
  $RCMAIL->kill_session();

  $auth = $RCMAIL->plugins->exec_hook('authenticate', array(
    'host' => $RCMAIL->autoselect_host(),
    'user' => trim(get_input_value('_user', RCUBE_INPUT_POST)),
    'pass' => get_input_value('_pass', RCUBE_INPUT_POST, true,
       $RCMAIL->config->get('password_charset', 'ISO-8859-1')),
    'cookiecheck' => true,
    'valid' => $request_valid,
  ));

  // check if client supports cookies
  if ($auth['cookiecheck'] && empty($_COOKIE)) {
    $OUTPUT->show_message("cookiesdisabled", 'warning');
  }
  else if ($auth['valid'] && !$auth['abort'] &&
        !empty($auth['host']) && !empty($auth['user']) &&
        $RCMAIL->login($auth['user'], $auth['pass'], $auth['host'])
  ) {
    // create new session ID, don't destroy the current session
    // it was destroyed already by $RCMAIL->kill_session() above
    $RCMAIL->session->remove('temp');
    $RCMAIL->session->regenerate_id(false);

    // send auth cookie if necessary
    $RCMAIL->authenticate_session();

    // log successful login
    rcmail_log_login();

    // restore original request parameters
    $query = array();
    if ($url = get_input_value('_url', RCUBE_INPUT_POST)) {
      parse_str($url, $query);

      // prevent endless looping on login page
      if ($query['_task'] == 'login')
        unset($query['_task']);
    }

    // allow plugins to control the redirect url after login success
    $redir = $RCMAIL->plugins->exec_hook('login_after', $query + array('_task' => 'mail'));
    unset($redir['abort']);

    // send redirect
    $OUTPUT->redirect($redir);
  }
  else {
    $error_code = is_object($IMAP) ? $IMAP->get_error_code() : -1;

    $OUTPUT->show_message($error_code < -1 ? 'imaperror' : (!$auth['valid'] ? 'invalidrequest' : 'loginfailed'), 'warning');
    $RCMAIL->plugins->exec_hook('login_failed', array(
      'code' => $error_code, 'host' => $auth['host'], 'user' => $auth['user']));
    $RCMAIL->kill_session();
  }
}

// end session (after optional referer check)
else if ($RCMAIL->task == 'logout' && isset($_SESSION['user_id']) && (!$RCMAIL->config->get('referer_check') || rcube_check_referer())) {
  $userdata = array('user' => $_SESSION['username'], 'host' => $_SESSION['imap_host'], 'lang' => $RCMAIL->user->language);
  $OUTPUT->show_message('loggedout');
  $RCMAIL->logout_actions();
  $RCMAIL->kill_session();
  $RCMAIL->plugins->exec_hook('logout_after', $userdata);
}

// check session and auth cookie
else if ($RCMAIL->task != 'login' && $_SESSION['user_id'] && $RCMAIL->action != 'send') {
  if (!$RCMAIL->authenticate_session()) {
    $OUTPUT->show_message('sessionerror', 'error');
    $RCMAIL->kill_session();
  }
}

// not logged in -> show login page
if (empty($RCMAIL->user->ID)) {
  if ($OUTPUT->ajax_call)
    $OUTPUT->redirect(array(), 2000);

  if (!empty($_REQUEST['_framed']))
    $OUTPUT->command('redirect', '?');

  // check if installer is still active
  if ($RCMAIL->config->get('enable_installer') && is_readable('./installer/index.php')) {
    $OUTPUT->add_footer(html::div(array('style' => "background:#ef9398; border:2px solid #dc5757; padding:0.5em; margin:2em auto; width:50em"),
      html::tag('h2', array('style' => "margin-top:0.2em"), "Installer script is still accessible") .
      html::p(null, "The install script of your Roundcube installation is still stored in its default location!") .
      html::p(null, "Please <b>remove</b> the whole <tt>installer</tt> folder from the Roundcube directory because .
        these files may expose sensitive configuration data like server passwords and encryption keys
        to the public. Make sure you cannot access the <a href=\"./installer/\">installer script</a> from your browser.")
      )
    );
  }

  $RCMAIL->set_task('login');
  $OUTPUT->send('login');
}
// CSRF prevention
else {
  // don't check for valid request tokens in these actions
  $request_check_whitelist = array('login'=>1, 'spell'=>1);

  // check client X-header to verify request origin
  if ($OUTPUT->ajax_call) {
    if (rc_request_header('X-Roundcube-Request') != $RCMAIL->get_request_token() && !$RCMAIL->config->get('devel_mode')) {
      header('HTTP/1.1 404 Not Found');
      die("Invalid Request");
    }
  }
  // check request token in POST form submissions
  else if (!empty($_POST) && !$request_check_whitelist[$RCMAIL->action] && !$RCMAIL->check_request()) {
    $OUTPUT->show_message('invalidrequest', 'error');
    $OUTPUT->send($RCMAIL->task);
  }

  // check referer if configured
  if (!$request_check_whitelist[$RCMAIL->action] && $RCMAIL->config->get('referer_check') && !rcube_check_referer()) {
    raise_error(array(
      'code' => 403,
      'type' => 'php',
      'message' => "Referer check failed"), true, true);
  }
}

// handle special actions
if ($RCMAIL->action == 'keep-alive') {
  $OUTPUT->reset();
  $OUTPUT->send();
}
else if ($RCMAIL->action == 'save-pref') {
  include 'steps/utils/save_pref.inc';
}


// map task/action to a certain include file
$action_map = array(
  'mail' => array(
    'preview' => 'show.inc',
    'print'   => 'show.inc',
    'moveto'  => 'move_del.inc',
    'delete'  => 'move_del.inc',
    'send'    => 'sendmail.inc',
    'expunge' => 'folders.inc',
    'purge'   => 'folders.inc',
    'remove-attachment'  => 'attachments.inc',
    'display-attachment' => 'attachments.inc',
    'upload' => 'attachments.inc',
    'group-expand' => 'autocomplete.inc',
  ),
  
  'addressbook' => array(
    'add' => 'edit.inc',
    'group-create' => 'groups.inc',
    'group-rename' => 'groups.inc',
    'group-delete' => 'groups.inc',
    'group-addmembers' => 'groups.inc',
    'group-delmembers' => 'groups.inc',
  ),

  'settings' => array(
    'folders'       => 'folders.inc',
    'rename-folder' => 'folders.inc',
    'delete-folder' => 'folders.inc',
    'subscribe'     => 'folders.inc',
    'unsubscribe'   => 'folders.inc',
    'purge'         => 'folders.inc',
    'folder-size'   => 'folders.inc',
    'add-identity'  => 'edit_identity.inc',
  )
);

// include task specific functions
if (is_file($incfile = 'program/steps/'.$RCMAIL->task.'/func.inc'))
  include_once($incfile);

// allow 5 "redirects" to another action
$redirects = 0; $incstep = null;
while ($redirects < 5) {
  $stepfile = !empty($action_map[$RCMAIL->task][$RCMAIL->action]) ?
    $action_map[$RCMAIL->task][$RCMAIL->action] : strtr($RCMAIL->action, '-', '_') . '.inc';
    
  // execute a plugin action
  if ($RCMAIL->plugins->is_plugin_task($RCMAIL->task)) {
    $RCMAIL->plugins->exec_action($RCMAIL->task.'.'.$RCMAIL->action);
    break;
  }
  else if (preg_match('/^plugin\./', $RCMAIL->action)) {
    $RCMAIL->plugins->exec_action($RCMAIL->action);
    break;
  }
  // try to include the step file
  else if (is_file($incfile = 'program/steps/'.$RCMAIL->task.'/'.$stepfile)) {
    include($incfile);
    $redirects++;
  }
  else {
    break;
  }
}


// parse main template (default)
$OUTPUT->send($RCMAIL->task);


// if we arrive here, something went wrong
raise_error(array(
  'code' => 404,
  'type' => 'php',
  'line' => __LINE__,
  'file' => __FILE__,
  'message' => "Invalid request"), true, true);

